File: python/src/grid_generator.py
import numpy as np

# Function to calculate the boundaries of the area based on central coordinates and distance
def calculate_boundaries(lat, lon, distance):
    meters_in_degree_lat = 111320  # Conversion factor from meters to degrees for latitude
    meters_in_degree_lon = 111320 * np.cos(np.radians(lat))  # Conversion factor for longitude, adjusted for latitude

    # Calculate the latitude and longitude increments for the given distance
    delta_lat = distance / meters_in_degree_lat
    delta_lon = distance / meters_in_degree_lon

    # Return the coordinates for the square's boundaries
    return lat - delta_lat, lon - delta_lon, lat + delta_lat, lon + delta_lon

# Function to create a grid within given latitude and longitude boundaries
def create_grid(lat_start, lon_start, lat_end, lon_end, step_meters):
    meters_in_degree_lat = 111320  # Conversion factor from meters to degrees for latitude
    meters_in_degree_lon = 111320 * np.cos(np.radians(lat_start))  # Conversion factor for longitude, adjusted for latitude

    # Calculate the step size in degrees for each grid square
    step_degrees_lat = step_meters / meters_in_degree_lat
    step_degrees_lon = step_meters / meters_in_degree_lon

    # Create arrays of latitude and longitude values using numpy
    latitudes = np.arange(lat_start, lat_end, step_degrees_lat)
    longitudes = np.arange(lon_start, lon_end, step_degrees_lon)

    # Generate grid points (latitude, longitude pairs)
    grid_points = [(lat, lon) for lat in latitudes for lon in longitudes]
    return grid_points

# Central geolocation coordinates
central_lat, central_lon = 52.392926, 13.092025
print(f"Central Geolocation: {central_lat}, {central_lon}")

# Define the boundaries of the area (4000m x 4000m area, 2000 meters from the center)
lat_start, lon_start, lat_end, lon_end = calculate_boundaries(central_lat, central_lon, 2000)
print(f"Area Boundaries:\nLatitude Start: {lat_start}\nLatitude End: {lat_end}\nLongitude Start: {lon_start}\nLongitude End: {lon_end}")

# Create the grid with 500x500 meters numbered squares
grid_points = create_grid(lat_start, lon_start, lat_end, lon_end, 500)
numbered_grid_points = {i: grid_points[i] for i in range(len(grid_points))}
print(f"Number of Grid Points: {len(numbered_grid_points)}")

# Print grid points with numbers
for number, point in numbered_grid_points.items():
    print(f"Square {number}: {point}")

File: react/src/gridGenerator.js
// Imports from geohash: Geohash is a public domain geocode system invented in 2008 by Gustavo Niemeyer and (similar work in 1966) G. Schimpf. It is a hierarchical spatial data structure which subdivides space into buckets of grid shape, which is one of the many applications of what is known as a Z-order curve, and generally space-filling curves.
// ngGeohash is a geohash library for AngularJS. It provides a service and a filter.
// Importing geohash library
import geohash from 'ngeohash';

// Function to calculate the bounds of a square based on its center and step sizes
function calculateSquareBounds(centerLat, centerLon, latStep, lonStep) {
    const halfLat = latStep / 2;
    const halfLon = lonStep / 2;

    // Calculate the start and end points for latitude and longitude
    const latStart = centerLat - halfLat;
    const lonStart = centerLon - halfLon;
    const latEnd = centerLat + halfLat;
    const lonEnd = centerLon + halfLon;
    
    // Return the calculated bounds
    return { latStart, lonStart, latEnd, lonEnd };
}

// Function to create a grid of geohashes within given latitude and longitude bounds
function createGridWithGeohash(latStart, lonStart, latEnd, lonEnd, stepMeters) {
    // Conversion factor from meters to degrees latitude (constant)
    const metersInDegreeLat = 111320;

    // Arrays to store calculated latitudes and longitudes
    let latitudes = [];
    let longitudes = [];
    // Array to store grid points with geohash
    let gridPointsWithGeohash = [];

    // Calculate step size in degrees for latitude
    const stepDegreesLat = stepMeters / metersInDegreeLat;

    // Populate latitudes array within the specified range
    for (let lat = latStart; lat <= latEnd; lat += stepDegreesLat) {
        latitudes.push(lat);
    }

    // For each latitude, calculate corresponding longitude step size
    latitudes.forEach(lat => {
        // Convert latitude from degrees to radians for accurate calculation
        const radiansLat = (lat * Math.PI) / 180;
        // Conversion factor from meters to degrees longitude, adjusted for this latitude
        const metersInDegreeLon = metersInDegreeLat * Math.cos(radiansLat);
        // Calculate step size in degrees for longitude
        const stepDegreesLon = stepMeters / metersInDegreeLon;

        // Reset longitudes array for each latitude
        longitudes = [];

        // Populate longitudes array within the specified range
        for (let lon = lonStart; lon <= lonEnd; lon += stepDegreesLon) {
            longitudes.push(lon);
        }

        // Combine latitudes and longitudes to create grid points
        longitudes.forEach(lon => {
            // Calculate the center point of each grid square
            const centerLat = lat + stepDegreesLat / 2;
            const centerLon = lon + stepDegreesLon / 2;

            // Generate geohash for the center point
            const geohashKey = geohash.encode(centerLat, centerLon);

            // Calculate bounds for the square
            const squareBounds = calculateSquareBounds(centerLat, centerLon, stepDegreesLat, stepDegreesLon);

            // Add the grid point with its geohash and bounds to the array
            gridPointsWithGeohash.push({ lat: centerLat, lon: centerLon, geohash: geohashKey, bounds: squareBounds });
        });
    });

    // Return the array of grid points with their corresponding geohashes
    return gridPointsWithGeohash;
}

// Export the function for use in other modules
export { createGridWithGeohash };
